server/src/test/java/com/angelomelonas/grpcwebchat/integration/ChatServiceTest.java

```diff
@@ -0,0 +1,260 @@
+
+import com.angelomelonas.grpcwebchat.Chat.AuthenticationRequest;
+import com.angelomelonas.grpcwebchat.Chat.AuthenticationResponse;
+import com.angelomelonas.grpcwebchat.Chat.Message;
+import com.angelomelonas.grpcwebchat.Chat.MessageRequest;
+import com.angelomelonas.grpcwebchat.Chat.SubscriptionRequest;
+import com.angelomelonas.grpcwebchat.Chat.UnsubscriptionRequest;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class ChatServiceTest {
+
+    private ChatServiceTestClient chatServiceTestClient;
+
+    @Before
+    public void resetChatClient() {
+        chatServiceTestClient = new ChatServiceTestClient();
+    }
+
+    @After
+    public void cleanUp() throws InterruptedException {
+        chatServiceTestClient.shutdown();
+    }
+
+    @Test
+    public void authenticateTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        Assert.assertNotNull(authenticationResponse.getUuid());
+    }
+
+    @Test
+    public void subscribeTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        UUID uuid = UUID.fromString(authenticationResponse.getUuid());
+        String username = "TestUsername" + randomSuffixGenerator();
+
+        SubscriptionRequest subscriptionRequest = SubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .setUsername(username)
+                .build();
+
+        StreamObserverTestHelper<Message> streamObserver = new StreamObserverTestHelper();
+
+        chatServiceTestClient.subscribe(subscriptionRequest, streamObserver);
+        // Wait for the server to acknowledge the subscription.
+        Message messageResponse = streamObserver.waitForOnNext().get();
+
+        Assert.assertEquals("Server", messageResponse.getUsername());
+        Assert.assertEquals("Client subscribed successfully.", messageResponse.getMessage());
+
+        UnsubscriptionRequest unsubscriptionRequest = UnsubscriptionRequest.newBuilder()
+                .setUuid(String.valueOf(uuid))
+                .build();
+
+        chatServiceTestClient.unsubscribe(unsubscriptionRequest);
+    }
+
+    @Test
+    public void unsubscribeTest() {
+        AuthenticationResponse authenticationResponse = chatServiceTestClient.authenticate(AuthenticationRequest.newBuilder().build());
+
+        UUID uuid = UUID.fromString(authenticationResponse.getUuid());
+        String username = "TestUsername" + randomSuffixGenerator();
```

Comment:
I don't think the client should send the username on subscribe it should be done in the 'AuthenticationRequest' request
and then the server will send back an UUID (like it does currently). There after the client no longer has to specify the username each time when talking to the server

Reply:
This is a design change I will consider (strongly) after all the other issues have been resolved :)